[ADD] New module 'mrp_production_hierarchy'

sebalix opened the pull request:
Production Orders Hierarchy
===========================

This module allows to view the hierarchy of generated production orders
when several levels of bill of materials are involved.

Usage
=====

Create a new production order (Manufacture / Operations / Manufacturing Orders)
with a multi-levels nomenclatured product, and related products configured with
'Manufacture' and 'Make To Order' routes.

Once saved, click on the 'Hierarchy' button at right.

Known issues / Roadmap
======================

* Only new production orders will benefit from this feature, it is not compatible with existing orders.

sebalix committed:
[ADD] New module 'mrp_production_hierarchy'

pedrobaeza commented:
Maybe a screenshot in the README will help to understand and showcase how this works.

sebalix commented:
@pedrobaeza there are already 3 screenshots?

pedrobaeza commented:
Sorry, I didn't see them!

JordiBForgeFlow commented:
@sebalix seems that you just show the first level of child, not the full hierarchy. Is that correct?

sebalix commented:
@jbeficent exactly, the parent MO is not shown here, but it is feasible if you prefer?
EDIT : except the parent MO, the full hierarchy is displayed, it's just that the BoM used (standard demo data) for the screenshot has only 1 level...

JordiBForgeFlow commented:
Ok! Thx

sebalix committed:
[FIX] Display the hierarchy with the parent MO included + New screenshots

sebalix commented:
Hierarchy treeview now includes the parent MO + Screenshots updated (the last one now shows a real hierarchy).

sebalix committed:
[IMP] mrp_production_hierarchy: Track the root production order on each generated sub-production orders + Display root order and direct parent order on the form

JordiBForgeFlow commented:
Yes, I see 
![image](https://user-images.githubusercontent.com/7683926/40244500-4e17376a-5ac3-11e8-8ac5-fcfb91c9a317.png)

sebalix committed:
[IMP] mrp_production_hierarchy: Only display root and parent production orders on the form if they exist

sebalix commented:
The module now displays the root and parent manufacturing order on a child's form.
I'm done with this PR, ready to be reviewed.

JordiBForgeFlow commented on mrp_production_hierarchy/models/mrp_production.py:
```diff
@@ -0,0 +1,55 @@
+    parent_right = fields.Integer('Right Parent', index=True)
+
+    @api.multi
+    def _generate_moves(self):
```
This is an api.multi. So self is likely to be a recordset, and in the  original method it iterates on that recordset.

IMHO you should iterate through the recordset,like:

for mo in self:
   if not mo.env.context.get('root_mrp_production_id'):
            mo = mo.with_context(root_mrp_production_id=mo.id)
   mo = mo.with_context(parent_mrp_production_id=mo.id)
   return super(MrpProduction, mo)._generate_moves()

sebalix replied:
Thanks, fixed.

sebalix committed:
[FIX] mrp_production_hierarchy: Fix method '_generate_moves()'

sebalix committed:
[FIX] mrp_production_hierarchy: Increase test coverage

sebalix committed:
[IMP] mrp_production_hierarchy: FR translations updated

sebalix committed:
[IMP] mrp_production_hierarchy: display root and parent order on the tree view + allow to group by the production orders on these fields

sebalix committed:
[IMP] mrp_production_hierarchy: do not display canceled production orders

sebalix committed:
[FIX] mrp_production_hierarchy: do not display canceled production orders

sebalix committed:
[FIX] mrp_production_hierarchy: disable _parent_store mechanism (MPTT) to fix issues when using the `child_of` search operator.

`(parent_left, parent_right)` duplicates can occur when creating multiple
manufacturing orders in parallel, which mess up the result of the `child_of`
search operator (wrong result returned).
Related Odoo issue: https://github.com/odoo/odoo/pull/14515
Odoo fix (for `stock.quant.package` here): https://github.com/odoo/odoo/pull/14694

This issue should not be relevant anymore in 11.0 as the `child_of/parent_of`
mechanism has been refactored (https://github.com/odoo/odoo/pull/22558)

sebalix commented:
We are experiencing issues with the `_parent_store =  True` setting on `mrp.production` when creating manufacturing orders concurrently: https://github.com/odoo/odoo/issues/14493
Solution by Odoo on 10.0 (related to `stock.quant.package` only): https://github.com/odoo/odoo/pull/14694
The commit b2e47e27ac497e189cfc8a9841d9b05581796d46 applies the same fix.

The `child_of` performance is related to the depth of the tree with `_parent_store = False`, and we are performing `child_of` search on MO trees with 400-500 children on 7-8 levels of depth without any problem.

sebalix committed:
[IMP] mrp_production_hierarchy: README

ovnicraft commented:
Here tested with 7 levels w/o problems.

LoisRForgeFlow reviewed (approved):
Functional review :+1:

remihb commented:
LGTM ;)
Module in production for 3 months, no issues were noted.
Need merge ?

LoisRForgeFlow commented:
@remihb next time do the review using github feature, so we can count your thumb up :)

/ocabot merge

LoisRForgeFlow commented:
Runbot issue has been fixed here: https://github.com/OCA/manufacture/pull/401

/ocabot merge
